### ISSUE TITLE
Cannot create array of QuantumCircuit instances

### ISSUE THREAD

**Issue by nonhermitian** [Qiskit/qiskit-terra#2059]
### Information

- **Qiskit Terra version**: master
- **Python version**:
- **Operating system**:

### What is the current behavior?
```python
q1 = QuantumRegister(1, 'q1')
q2 = QuantumRegister(2, 'q2')
q3 = QuantumRegister(3, 'q3')
qc = QuantumCircuit(q1, q2, q3)
np.asarray([qc])
```
```python
array([], shape=(1, 0), dtype=float64)
```

**Comment by 1ucian0** [comment]
```
import numpy as np
from numpy import array, float64
from qiskit import QuantumCircuit, QuantumRegister

q1 = QuantumRegister(1, 'q1')
q2 = QuantumRegister(2, 'q2')
q3 = QuantumRegister(3, 'q3')
qc = QuantumCircuit(q1, q2, q3)
np.asarray([qc])

array([], shape=(1, 0), dtype=float64)
```
```
Traceback (most recent call last):
  File "2059.py", line 11, in <module>
    array([], shape=(1, 0), dtype=float64)
TypeError: 'shape' is an invalid keyword argument for this function
```

What am I doing wrong?

**Comment by nonhermitian** [comment]
That is the output.

**Comment by mtreinish** [comment]
`array([], shape=(1, 0), dtype=float64)` is the output from the script not part of it. Remove that line and it will return that.

**Comment by mtreinish** [comment]
This is a duplicate of: https://github.com/Qiskit/qiskit-terra/issues/1898 so closing it to track it there

**Comment by mtreinish** [comment]
Err I guess it's slightly different #1898 is for an array of registers and this is for an array of circuits. I'll keep it separate for now and we can integrate if the solution ends up being the same

**Comment by nonhermitian** [comment]
It is different. in contrast to the title of #1898, you can create an array of registers:

```python
np.array([QuantumRegister(2)])
```
the output is a bit funky though:

```python
array([[[QuantumRegister(2, 'q1'), 0],
        [QuantumRegister(2, 'q1'), 1]]], dtype=object)
```

**Comment by 1ucian0** [comment]
And what's the expected output?

**Comment by nonhermitian** [comment]
A 1D array with elements that are `QuantumCircuit` instances:
```python
array([QuantumCircuit], shape=(1,), dtype=object)
```

**Comment by 1ucian0** [comment]
It seems that happens because `QuantumCircuit.__getitem__` pulls from `QuantumCircuit.data`.

**Comment by 1ucian0** [comment]
This issue seems fixed already in Terra 0.11.

```python
q1 = QuantumRegister(1, 'q1')
q2 = QuantumRegister(2, 'q2')
q3 = QuantumRegister(3, 'q3')
qc = QuantumCircuit(q1, q2, q3)
np.asarray([qc])
```
```
array([], shape=(1, 0), dtype=float64)
```

Feel free to repoen if that's not the case.
